docs/core/deploying/trimming/trim-warnings/il2122.md

```diff
@@ -0,0 +1,30 @@
+---
+title: "IL2122: Type 'type' is not assembly qualified. Type name strings used for dynamically accessing a type should be assembly qualified."
+description: "Learn about trim warning IL2122: TypeNameIsNotAssemblyQualified"
+ms.date: 08/23/24
+author: svbomer
+f1_keywords:
+  - "IL2122"
+---
+# IL2122: Type 'type' is not assembly qualified. Type name strings used for dynamically accessing a type should be assembly qualified
+
+## Cause
+
+Type name strings representing dynamically accessed types must be assembly qualified. Otherwise, the lookup semantics of `Type.GetType` will search the assembly with the `Type.GetType` callsite and the core library. The assembly with the `Type.GetType` callsite may be different than the assembly which passes the type name string to a location with `DynamicallyAccessedMembersAttribute`, so the tooling cannot determine which assemblies to search.
+
+## Example
+
+```csharp
+void TestInvalidTypeName()
+{
+    // IL2122: Type 'UnqualifiedTypeName' is not assembly qualified. Type name strings used for dynamically accessing a type should be assembly qualified.
+    RequirePublicMethodOnAType("UnqualifiedTypeName");
+}
+
+void RequirePublicMethodOnAType(
+    [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
+    string typeName)
+{
+    Type.GetType(typeName);
+}
```

Comment:
Maybe the example could show a scenario where `Type.GetType()` could return a different type? I'm fine to leave it if that might be too confusing, though.

```suggestion
// In Assembly 1
class MyClass
{
    void TestInvalidTypeName()
    {
        // IL2122: Type 'UnqualifiedTypeName' is not assembly qualified. Type name strings used for dynamically accessing a type should be assembly qualified.
        RequirePublicMethodOnAType("UnqualifiedTypeName");
    }

    void ForwardTypeNameToAnotherMethod(
        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
        string typeName)
    {
        MyTypeFromAnotherAssembly.GetType(typeName);
    }
}

// In Assembly 2
public class MyTypeFromAnotherAssembly
{
    void GetTypeAndSearchThroughMethods(
        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)]
        string typeName)
    {
        Type.GetType(typeName).GetMethods();
    }
}
```